File: Data Structures & Algorithms/course-schedule-ii/submission-0.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        preMap = {i:[] for i in range(numCourses)}
        for crs, pre in prerequisites:
            preMap[crs].append(pre)

        res = []
        cycle, visit = set(), set()
        
        def dfs(crs):
            if crs in visit:
                return True
            if crs in cycle:
                return False
            cycle.add(crs)
            for pre in preMap[crs]:
                if not dfs(pre):
                    return False
            cycle.remove(crs)
            visit.add(crs)
            res.append(crs)
            return True

        for i in range(numCourses):
            if dfs(i) == False: 
                return []
        return res
        

File: Data Structures & Algorithms/palindrome-partitioning/submission-0.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        List<String> part = new ArrayList<>();

        dfs(0,s,res,part);
        return res;
    }

    private void dfs(int i, String s, List<List<String>> res, List<String> part) {
        if(i >= s.length()) {
            res.add(new ArrayList<>(part));
        }
        for(int j=i; j<s.length(); j++) {
            if(isPal(s, i, j)) {
                part.add(s.substring(i, j+1));
                dfs(j+1, s, res, part);
                part.remove(part.size() - 1);
            }
        }
    }

    private boolean isPal(String s, int l, int r) {
        while(l<r) {
            if(s.charAt(l) != s.charAt(r)) {
                return false;
            }
            r--;
            l++;
        }
        return true;
    }
}
